Pre-register OAuth2 applications for git credential helpers

denyskon opened the pull request:
This PR is an extended implementation of #25189 and builds upon the proposal by @hickford in #25653, utilizing some ideas proposed internally by @wxiaoguang.

Mainly, this PR consists of a mechanism to pre-register OAuth2 applications on startup, which can be enabled or disabled by modifying the `[oauth2].DEFAULT_APPLICATIONS` parameter in app.ini. The OAuth2 applications registered this way are being marked as "locked" and neither be deleted nor edited over UI to prevent confusing/unexpected behavior. Instead, they're being removed if no longer enabled in config.

![grafik](https://github.com/go-gitea/gitea/assets/47871822/81a78b1c-4b68-40a7-9e99-c272ebb8f62e)

The implemented mechanism can also be used to pre-register other OAuth2 applications in the future, if wanted.

Co-authored-by: hickford <mirth.hickford@gmail.com>
Co-authored-by: wxiaoguang <wxiaoguang@gmail.com>

hickford reviewed (commented):
LGTM. Nice solution

lunny commented on models/auth/oauth2.go:
```diff
@@ -46,6 +48,64 @@ func init() {
+		},
+	}
+
+	registeredOAuth2Apps, _ := GetOAuth2ApplicationsByUserID(ctx, 0)
```
Since `client_id` is unique, searching by the `client_id` is the best resolution here. You can use `In` because the default applications should not too much. Gitea supports Global Applications, although there should not be too many global applications.

denyskon replied:
The idea behind that line is that we only retrieve the list once, as these OAuth2 apps are being registered globally. I can change it to a lookup by client_id, but that will increase the number of DB calls.....

denyskon replied:
If that's not a problem, I'd also prefer a client_id lookup

lunny replied:
You can use client_id IN (client_id1, client_id2) which is only 1 query.

denyskon replied:
Done

lunny commented on modules/util/slice.go:
```diff
@@ -88,3 +88,13 @@ func SliceRemoveAllFunc[T comparable](slice []T, targetFunc func(T) bool) []T {
 }
+
+// SliceFind returns the element id of the first element that matches the condition, otherwise -1
+func SliceFind[T any](slice []T, condFunc func(T) bool) int {
```
Maybe we can use `SliceContainsFunc` directly?

denyskon replied:
I don't think so, it only returns bool, while we need the specific element or its id.

lunny commented on models/migrations/v1_21/v271.go:
```diff
@@ -0,0 +1,15 @@
+	"xorm.io/xorm"
+)
+
+func OAuth2ApplicationAddLockedProperty(x *xorm.Engine) error {
```
You need to reference this function on `migrations.go`.

denyskon replied:
Oops, my first time creating a migration :laughing: Could've used the proposed migration docs from #23793

lunny commented:
Maybe we need to check the delete functions to don't allow delete OAuth2 Applications from both UI and API.

denyskon commented:
@lunny done

wxiaoguang commented:
Actually, the "ClientID" can be any string, so maybe it's even better to use "builtin-git-cred-oauth" / "builtin-git-cred-manager", it brings consistent user experience for the config too:

```
DEFAULT_APPLICATIONS = builtin-git-cred-oauth, builtin-git-cred-manager
```

And the pre-defined ClientIDs could avoid the database migration (no need to add the Locked field)


```
		{
			Name:         "Git Credential Manager",
			ClientID:     "builtin-git-cred-manager",
			RedirectURIs: []string{"http://127.0.0.1", "https://127.0.0.1"},
		},
```

wxiaoguang commented:
If you don't mind, could I push some changes for idea https://github.com/go-gitea/gitea/pull/26291#issuecomment-1664907425 ?

denyskon commented:
@wxiaoguang Fell free to make any changes if you want

wxiaoguang commented:
-> https://github.com/denyskon/gitea/pull/3

some changes:

1. Not use Locked field, then no need to do DB migration
2. Introduce "ConfigName" (for the config setting value), to avoid using spaces or other config-unfriendly characters for the option value

There is still a question: should it allow users to "edit" the builtin oauth2 applications?

![image](https://github.com/go-gitea/gitea/assets/2114189/ad049c3b-e77b-4abd-a272-12a68747d35f)

denyskon commented:
Thanks a lot @wxiaoguang, this new logic looks much cleaner

I think editing is okay, as admins might want to change the display name.

And about the two client IDs: I don't think it's a good idea to let two different apps use the same client ID.......

lunny commented:
> Thanks a lot @wxiaoguang, this new logic looks much cleaner
> 
> I think editing is okay, as admins might want to change the display name.
> 
> And about the two client IDs: I don't think it's a good idea to let two different apps use the same client ID.......

But once admin changed the name, when gitea restart, a new record will be created again?

denyskon commented:
We check the client ID, so the display name change doesn't have any effect on the registration.

wxiaoguang commented:
> And about the two client IDs: I don't think it's a good idea to let two different apps use the same client ID.......

I haven't got a clear idea for it. But I can't imagine a case for why using the only one shared ClientID is bad, either.

According to [Occam's razor](https://en.wikipedia.org/wiki/Occam%27s_razor): if adding more ClientIDs is not necessary, then only use one?

@hickford Do you have more thoughts about this question? Could these "git credential helpers" use the same/shared ClientID if there is no bad case?

hickford commented:
> why it needs two applications? If I understand correctly, the "git-credential-oauth" and "Git Credential Manager" just need a "ClientID" to work. So, maybe there could be only one shared application for both of them?

It's necessary to define two applications so that the user consent screen shows correct details of the expected application   https://imgur.com/a/7RRUPES . Users can revoke consent for each application independently.

> admins might want to change the display name

This is a bad idea because it would confuse users.

denyskon commented:
I think it should be possible to change the display name. For example I run a Gitea instance as part of a school infrastructure, and there are people in my team that might need to work with git, but don't know anything about it, so I try to give everything a human-friendly name.

I understand why you think that it could confuse users, but I think only those will change something who really know why they're doing it :)

denyskon commented:
@wxiaoguang Do you want any more changes to be made?

wxiaoguang reviewed (approved):
I think the client IDs could be written into the document, to make them could be found easily.

denyskon commented:
I changed the behavior to disallow editing and also changed the UI, as now we don't need any buttons (and as there is no proper way to disable link buttons :laughing:)

![grafik](https://github.com/go-gitea/gitea/assets/47871822/68180f2d-aa6b-4e2c-887b-91846480c850)
